raja@0.14.0 +rocm: rocprim: error: no template named 'enable_if_t' in namespace 'std'

### Steps to reproduce the issue

Having some trouble building `raja@0.14.0 +rocm`.

The details posted here apply to OLCF Crusher w/ externally-installed ROCm 5.2.0 (see spack.yaml below) but also when installing ROCm components with Spack (i.e. not external).

This same error was reported in another context here:
* https://github.com/ROCmSoftwarePlatform/rocPRIM/issues/349
* The recommended solution of specifying `-std=c++14` to `hipcc` resolved the issue for me.

I plan to submit a PR adding `-std=c++14` to `HIP_HIPCC_FLAGS`, but may need some help figuring out what versions of the rocPRIM/rocTHRUST require this.

spack.yaml: [spack.yaml.txt](https://github.com/spack/spack/files/9834858/spack.yaml.txt)

<details><summary>Concretization</summary><pre>
 -   raja@0.14.0%clang@14.0.0-rocm5.2.0~cuda+examples+exercises~ipo~openmp+rocm+shared~tests amdgpu_target=gfx90a build_type=RelWithDebInfo arch=cray-sles15-zen3
[+]      ^blt@0.4.1%clang@14.0.0-rocm5.2.0 arch=cray-sles15-zen3
[+]          ^cmake@3.20.6%clang@14.0.0-rocm5.2.0~doc+ncurses+ownlibs~qt build_type=Release arch=cray-sles15-zen3
[+]              ^ncurses@6.3%clang@14.0.0-rocm5.2.0~symlinks+termlib abi=none arch=cray-sles15-zen3
[+]                  ^pkgconf@1.8.0%clang@14.0.0-rocm5.2.0 arch=cray-sles15-zen3
[+]              ^openssl@1.1.1q%clang@14.0.0-rocm5.2.0~docs~shared certs=mozilla patches=3fdcf2d arch=cray-sles15-zen3
[+]                  ^ca-certificates-mozilla@2022-07-19%clang@14.0.0-rocm5.2.0 arch=cray-sles15-zen3
[+]                  ^perl@5.34.1%clang@14.0.0-rocm5.2.0+cpanm+shared+threads arch=cray-sles15-zen3
[+]                      ^berkeley-db@18.1.40%clang@14.0.0-rocm5.2.0+cxx~docs+stl patches=26090f4,b231fcc arch=cray-sles15-zen3
[+]                      ^bzip2@1.0.8%clang@14.0.0-rocm5.2.0~debug~pic+shared arch=cray-sles15-zen3
[+]                          ^diffutils@3.8%clang@14.0.0-rocm5.2.0 arch=cray-sles15-zen3
[+]                              ^libiconv@1.16%clang@14.0.0-rocm5.2.0 libs=shared,static arch=cray-sles15-zen3
[+]                      ^gdbm@1.19%clang@14.0.0-rocm5.2.0 arch=cray-sles15-zen3
[+]                          ^readline@8.1.2%clang@14.0.0-rocm5.2.0 arch=cray-sles15-zen3
[+]                      ^zlib@1.2.12%clang@14.0.0-rocm5.2.0+optimize+pic+shared patches=0d38234 arch=cray-sles15-zen3
[+]      ^camp@0.2.3%clang@14.0.0-rocm5.2.0~cuda~ipo~openmp+rocm~tests amdgpu_target=gfx90a build_type=RelWithDebInfo arch=cray-sles15-zen3
[+]          ^hip@5.2.0%clang@14.0.0-rocm5.2.0~ipo build_type=Release patches=959d1fe arch=cray-sles15-zen3
[+]          ^hsa-rocr-dev@5.2.0%clang@14.0.0-rocm5.2.0+image~ipo+shared build_type=Release patches=71e6851 arch=cray-sles15-zen3
[+]          ^llvm-amdgpu@5.2.0%clang@14.0.0-rocm5.2.0~ipo~link_llvm_dylib~llvm_dylib~openmp+rocm-device-libs build_type=Release patches=a08bbe1 arch=cray-sles15-zen3
[+]      ^rocprim@5.2.0%clang@14.0.0-rocm5.2.0~ipo amdgpu_target=auto build_type=Release arch=cray-sles15-zen3
</pre></details>


Build error:
```
$> spack -e . install
...
==> Installing raja-0.14.0-nqqe5i7nb7yb3rrny4n65zvqplkrbfwd
==> No binary for raja-0.14.0-nqqe5i7nb7yb3rrny4n65zvqplkrbfwd found: installing from source
==> Warning: Fetching from mirror without a checksum!
  This package is normally checked out from a version control system, but it has been archived on a spack mirror.  This means we cannot know a checksum for the tarball in advance. Be sure that your connection to this mirror is secure!
==> Using cached archive: /gpfs/alpine/csc439/world-shared/E4S/ParaTools/exago/ci/spack/var/spack/cache/_source-cache/git//LLNL/RAJA.git/v0.14.0.tar.gz
==> No patches needed for raja
==> raja: Executing phase: 'initconfig'
==> raja: Executing phase: 'cmake'
==> raja: Executing phase: 'build'
...
/opt/rocm-5.2.0/bin/hipcc -c /gpfs/alpine/csc439/world-shared/E4S/ParaTools/exago/ci/build_stage/eugeneswalker/spack-stage-raja-0.14.0-nqqe5i7nb7yb3rrny4n65zvqplkrbfwd/spack-src/examples/tut_tiled-matrix-transpose.cpp -o /gpfs/alpine/csc439/world-shared/E4S/ParaTools/exago/ci/build_stage/eugeneswalker/spack-stage-raja-0.14.0-nqqe5i7nb7yb3rrny4n65zvqplkrbfwd/spack-build-nqqe5i7/examples/CMakeFiles/tut_tiled-matrix-transpose.dir//./tut_tiled-matrix-transpose_generated_tut_tiled-matrix-transpose.cpp.o --amdgpu-target=gfx90a -D__HIP_PLATFORM_HCC__ -D__HIP_ROCclr__ -D__HIP_PLATFORM_AMD__ -D__HIP_PLATFORM_HCC__ -D__HIP_ROCclr__ -D__HIP_PLATFORM_AMD__ -fPIC -fPIC -g -O2 -I/gpfs/alpine/csc439/world-shared/E4S/ParaTools/exago/ci/build_stage/eugeneswalker/spack-stage-raja-0.14.0-nqqe5i7nb7yb3rrny4n65zvqplkrbfwd/spack-src/include -I/gpfs/alpine/csc439/world-shared/E4S/ParaTools/exago/ci/build_stage/eugeneswalker/spack-stage-raja-0.14.0-nqqe5i7nb7yb3rrny4n65zvqplkrbfwd/spack-build-nqqe5i7/include -I/opt/rocm-5.2.0/include -I/opt/rocm-5.2.0/include -I/opt/rocm-5.2.0/include -I/opt/rocm-5.2.0/include -I/gpfs/alpine/csc439/world-shared/E4S/ParaTools/exago/ci/spack/opt/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placehold/cray-sles15-zen3/clang-14.0.0-rocm5.2.0/camp-0.2.3-o6lso5nnsh5l5rihceidkssqflka2xia/include
In file included from /gpfs/alpine/csc439/world-shared/E4S/ParaTools/exago/ci/build_stage/eugeneswalker/spack-stage-raja-0.14.0-nqqe5i7nb7yb3rrny4n65zvqplkrbfwd/spack-src/examples/resource-kernel.cpp:9:
In file included from /gpfs/alpine/csc439/world-shared/E4S/ParaTools/exago/ci/build_stage/eugeneswalker/spack-stage-raja-0.14.0-nqqe5i7nb7yb3rrny4n65zvqplkrbfwd/spack-src/include/RAJA/RAJA.hpp:73:
In file included from /gpfs/alpine/csc439/world-shared/E4S/ParaTools/exago/ci/build_stage/eugeneswalker/spack-stage-raja-0.14.0-nqqe5i7nb7yb3rrny4n65zvqplkrbfwd/spack-src/include/RAJA/policy/hip.hpp:33:
In file included from /gpfs/alpine/csc439/world-shared/E4S/ParaTools/exago/ci/build_stage/eugeneswalker/spack-stage-raja-0.14.0-nqqe5i7nb7yb3rrny4n65zvqplkrbfwd/spack-src/include/RAJA/policy/hip/scan.hpp:30:
In file included from /opt/rocm-5.2.0/include/rocprim/device/device_scan.hpp:31:
/opt/rocm-5.2.0/include/rocprim/device/../detail/various.hpp:60:5: error: no template named 'enable_if_t' in namespace 'std'; did you mean 'Enable_if_t'?
    std::enable_if_t<::rocprim::is_integral<T>::value && ::rocprim::is_integral<U>::value, int> = 0>
    ^~~~~
/opt/rocm-5.2.0/hip/include/hip/amd_detail/../../../../include/hip/amd_detail/amd_hip_fp16.h:77:9: note: 'Enable_if_t' declared here
        using Enable_if_t = typename std::enable_if<cond, T>::type;
        ^
...
     1319    In file included from /opt/rocm-5.2.0/include/rocprim/device/device_scan.hpp:33:
  >> 1320    /opt/rocm-5.2.0/include/rocprim/device/device_scan_config.hpp:83:9: error: no matching function for call to 'ceiling_div'
     1321            ::rocprim::detail::ceiling_div<unsigned int>(sizeof(Value), sizeof(int));
     1322            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  >> 1323    /opt/rocm-5.2.0/include/rocprim/device/device_scan_config.hpp:99:9: error: no matching function for call to 'ceiling_div'
     1324            ::rocprim::detail::ceiling_div<unsigned int>(sizeof(Value), sizeof(int));
     1325            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  >> 1326    /opt/rocm-5.2.0/include/rocprim/device/device_scan_config.hpp:116:9: error: no matching function for call to 'ceiling_div'
     1327            ::rocprim::detail::ceiling_div<unsigned int>(sizeof(Value), sizeof(int));
     1328            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  >> 1329    /opt/rocm-5.2.0/include/rocprim/device/device_scan_config.hpp:132:9: error: no matching function for call to 'ceiling_div'
     1330            ::rocprim::detail::ceiling_div<unsigned int>(sizeof(Value), sizeof(int));
     1331            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  >> 1332    /opt/rocm-5.2.0/include/rocprim/device/device_scan_config.hpp:149:9: error: no matching function for call to 'ceiling_div'
     1333            ::rocprim::detail::ceiling_div<unsigned int>(sizeof(Value), sizeof(int));
...
     1352    In file included from /opt/rocm-5.2.0/include/rocprim/device/device_scan.hpp:31:
  >> 1353    /opt/rocm-5.2.0/include/rocprim/device/../detail/various.hpp:60:5: error: no template named 'enable_if_t' in namespace 'std'; did you mean 'Enable_if_t'?
     1354        std::enable_if_t<::rocprim::is_integral<T>::value && ::rocprim::is_integral<U>::value, int> = 0>
     1355        ^~~~~
     1356    /opt/rocm-5.2.0/hip/include/hip/amd_detail/../../../../include/hip/amd_detail/amd_hip_fp16.h:77:9: note: 'Enable_if_t' declared here
     1357            using Enable_if_t = typename std::enable_if<cond, T>::type;
     1358            ^
  >> 1359    /opt/rocm-5.2.0/include/rocprim/device/device_scan_config.hpp:116:9: error: no matching function for call to 'ceiling_div'
     1360            ::rocprim::detail::ceiling_div<unsigned int>(sizeof(Value), sizeof(int));
     1361            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
...
```


### Error message

<details><summary>Error message</summary><pre>
... see above
</pre></details>


### Information on your system

* **Spack:** 0.19.0.dev0 (9621b2bb3b226cc887c2a7a5b04e85e24c466f1a)
* **Python:** 3.9.7
* **Platform:** cray-sles15-zen3
* **Concretizer:** clingo

### Additional information

[spack-build-out.txt](https://github.com/spack/spack/files/9834852/spack-build-out.txt)
[spack-build-env.txt](https://github.com/spack/spack/files/9834853/spack-build-env.txt)
[CMakeOutput.log](https://github.com/spack/spack/files/9834854/CMakeOutput.log)
[CMakeError.log](https://github.com/spack/spack/files/9834855/CMakeError.log)

FYI @wspear @davidbeckingsale @CameronRutherford 

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/raja/package.py
+++ b/var/spack/repos/builtin/packages/raja/package.py
@@ -133,12 +133,14 @@ class Raja(CachedCMakePackage, CudaPackage, ROCmPackage):
             entries.append(cmake_cache_option("ENABLE_HIP", True))
             entries.append(cmake_cache_path("HIP_ROOT_DIR", "{0}".format(spec["hip"].prefix)))
             hip_repair_cache(entries, spec)
+            hipcc_flags = []
+            if self.spec.satisfies("@0.14.0"):
+                hipcc_flags.append("-std=c++14")
             archs = self.spec.variants["amdgpu_target"].value
             if archs != "none":
                 arch_str = ",".join(archs)
-                entries.append(
-                    cmake_cache_string("HIP_HIPCC_FLAGS", "--amdgpu-target={0}".format(arch_str))
-                )
+                hipcc_flags.append("--amdgpu-target={0}".format(arch_str))
+            entries.append(cmake_cache_string("HIP_HIPCC_FLAGS", " ".join(hipcc_flags)))
         else:
             entries.append(cmake_cache_option("ENABLE_HIP", False))
 

